File: algorithms/Counting Sort/C++/code.cpp
#include <iostream>
#include <vector>

void countingSort(std::vector<int>& arr)
{
    if (arr.size() <= 1)
    {
        return;
    }

    int max = arr[0];

    for (int value : arr)
    {
        if (value > max)
        {
            max = value;
        }
    }

    std::vector<int> count(max + 1, 0);

    for (int value : arr)
    {
        count[value]++;
    }

    int index = 0;

    for (int value = 0; value <= max; value++)
    {
        while (count[value] > 0)
        {
            arr[index] = value;
            index++;
            count[value]--;
        }
    }
}

int main()
{
    std::vector<int> arr = {4, 2, 2, 8, 3, 3, 1};

    countingSort(arr);

    for (int value : arr)
    {
        std::cout << value << " ";
    }

    std::cout << "\n";

    return 0;
}

File: algorithms/Counting Sort/Python/code.py
def counting_sort(arr):
    if len(arr) <= 1:
        return

    max_value = arr[0]

    for value in arr:
        if value > max_value:
            max_value = value

    count = [0] * (max_value + 1)

    for value in arr:
        count[value] += 1

    index = 0

    for value in range(max_value + 1):
        while count[value] > 0:
            arr[index] = value
            index += 1
            count[value] -= 1


arr = [4, 2, 2, 8, 3, 3, 1]

counting_sort(arr)

print(arr)